File: src/main/java/com/zl/mjga/platform/events/InvocationContext.java
package com.zl.mjga.platform.events;

import java.util.Objects;
import java.util.Optional;

/** Trusted framework-neutral identity and trace context propagated across Capability Interfaces. */
public record InvocationContext(
    String actor,
    Optional<String> tenant,
    String correlationId,
    Optional<String> causationId,
    InvocationKind kind) {

  public InvocationContext {
    actor = requireText(actor, "actor");
    Objects.requireNonNull(tenant, "tenant");
    correlationId = requireText(correlationId, "correlationId");
    Objects.requireNonNull(causationId, "causationId");
    Objects.requireNonNull(kind, "kind");
  }

  private static String requireText(String value, String name) {
    if (value.isBlank()) {
      throw new IllegalArgumentException(name + " must not be blank");
    }
    return value;
  }

  public enum InvocationKind {
    USER,
    SYSTEM,
    SCHEDULED,
    EVENT
  }
}

File: src/main/kotlin/com/zl/mjga/platform/events/InvocationContext.kt
package com.zl.mjga.platform.events

/** Trusted framework-neutral identity and trace context propagated across Capability Interfaces. */
@org.springframework.modulith.NamedInterface("api")
data class InvocationContext(
    val actor: String,
    val tenant: String?,
    val correlationId: String,
    val causationId: String?,
    val kind: InvocationKind,
) {
    init {
        require(actor.isNotBlank()) { "actor must not be blank" }
        require(correlationId.isNotBlank()) { "correlationId must not be blank" }
    }

    enum class InvocationKind { USER, SYSTEM, SCHEDULED, EVENT }
}
